File: 6/solution.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
)

func countUnique(lines []string) int {
	letters := make(map[string]struct{})
	sum := 0
	for _, line := range lines {
		if len(line) == 0 {
			sum += len(letters)
			letters = make(map[string]struct{})
			continue
		}
		for _, letter := range line {
			letters[string(letter)] = struct{}{}
		}
	}
	// For the last group in input
	sum += len(letters)
	return sum
}

func intersection(set1 map[string]bool, set2 map[string]bool) map[string]bool {
	resultingIntersection := make(map[string]bool)
	for k := range set1 {
		if set2[k] {
			resultingIntersection[k] = true
		}
	}
	return resultingIntersection
}

func groupIntersection(group []map[string]bool) map[string]bool {
	result := group[0]
	for _, line := range group[1:] {
		result = intersection(result, line)
	}
	return result
}

func countCommon(lines []string) int {
	sum := 0
	var lineGroup []map[string]bool
	for _, line := range lines {
		if len(line) == 0 {
			sum += len(groupIntersection(lineGroup))
			lineGroup = nil
			continue
		}

		letters := make(map[string]bool)
		for _, letter := range line {
			letters[string(letter)] = true
		}
		lineGroup = append(lineGroup, letters)
	}
	// For the last group in input
	sum += len(groupIntersection(lineGroup))
	return sum

}

func part1(lines []string) int {
	return countUnique(lines)
}

func part2(lines []string) int {
	return countCommon(lines)
}

func main() {
	flag.Parse()
	if len(flag.Args()) == 0 {
		log.Fatal("Provide the path to input")
	}
	file, err := os.Open(flag.Args()[0])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var lines []string
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}

	fmt.Printf("Part 1 answer:%v\nPart 2 answer:%v\n", part1(lines), part2(lines))
}

File: 6/solution.py
import sys


def count_unique(lines):
    letters = set()
    count = 0
    for line in lines:
        if line == "\n":
            count += len(letters)
            letters = set()
            continue

        for letter in line:
            if letter != '\n':
                letters.add(letter)
    count += len(letters)
    return count


def part1(lines):
    return count_unique(lines)


def count_common(lines):
    line_group = []
    letters = set()
    count = 0
    for line in lines:
        if line == "\n":
            count += len(line_group[0].intersection(*line_group[1:]))
            line_group = []
            continue
        letters = set()
        for letter in line:
            if letter != '\n':
                letters.add(letter)
        line_group.append(letters)
    count += len(line_group[0].intersection(*line_group[1:]))
    return count


def part2(lines):
    return count_common(lines)


try:
    f = open(sys.argv[1], "r")
except IOError:
    print("Error opening the file, try again")
    sys.exit(1)
with f:
    rawLines = f.readlines()
    f.close()
    # digitArray = [int(s) for s in lines]
    print(
        f"Part 1 answer:{part1(rawLines)} Part 2 answer: {part2(rawLines)}")
